aws/credentials/processcreds/provider.go

```diff
@@ -0,0 +1,353 @@
+			"missing AccessKeyId in credential_process output", nil)
+	}
+
+	if len(resp.SecretAccessKey) == 0 {
+		return credentials.Value{ProviderName: ProviderName}, awserr.New(ErrCodeProcessProviderRequired,
+			"missing SecretAccessKey in credential_process output", nil)
+	}
+
+	// Handle expiration
+	if resp.Expiration != nil {
+		p.SetExpiration(*resp.Expiration, p.ExpiryWindow)
+	} else {
+		p.staticCreds = true
+	}
+
+	return credentials.Value{
+		ProviderName:    ProviderName,
+		AccessKeyID:     resp.AccessKeyID,
+		SecretAccessKey: resp.SecretAccessKey,
+		SessionToken:    resp.SessionToken,
+	}, nil
+}
+
+// IsExpired returns true if the credentials retrieved are expired, or not yet
+// retrieved.
+func (p *ProcessProvider) IsExpired() bool {
+	if p.staticCreds {
+		return false
+	}
+	return p.Expiry.IsExpired()
+}
+
+// LimitedBuffer is a buffer wrapper that only allows a max amount of data.
+type LimitedBuffer struct {
+	maxSize int
+	buff    *bytes.Buffer
+}
+
+// Write to the LimitedBuffer
+func (b *LimitedBuffer) Write(p []byte) (int, error) {
+	if len(p)+b.buff.Len() > b.maxSize {
+		return -1, fmt.Errorf("buffer size (%v) exceeded: %v", b.maxSize, len(p)+b.buff.Len())
+	}
+	b.buff.Write(p)
+	return len(p), nil
+}
+
+// Bytes retrieves data from the LimitedBuffer
+func (b *LimitedBuffer) Bytes() []byte {
+	return b.buff.Bytes()
+}
+
+// NewBuffer creates a new limited buffer.
+func NewBuffer(start, max int) *LimitedBuffer {
+	return &LimitedBuffer{
+		maxSize: max,
+		buff:    bytes.NewBuffer(make([]byte, 0, start)),
+	}
+}
+
+// executeCredentialProcess executes the `process` command on the OS and
+// returns the results or an error.
+func (p *ProcessProvider) executeCredentialProcess() ([]byte, error) {
+	command := strings.TrimSpace(p.Process)
+	if command == "" {
+		return nil, fmt.Errorf("process must be a non-empty string")
+	}
+
+	var cmdargs []string
+	var env []string
+	if runtime.GOOS == "windows" {
+		if comspec, ok := os.LookupEnv("ComSpec"); ok {
+			cmdargs = []string{comspec, "/C"}
+		} else {
+			cmdargs = []string{"cmd", "/C"}
+		}
+		if _, ok := os.LookupEnv("PATH"); !ok {
+			env = append(env, fmt.Sprintf("%s=%s", "PATH", "C:\\Windows\\system32"))
+		}
+
+	} else {
+		cmdargs = []string{"/bin/sh", "-c"}
+		if _, ok := os.LookupEnv("PATH"); !ok {
+			env = append(env, fmt.Sprintf("%s=%s", "PATH", "/bin"))
+		}
+	}
+	cmdargs = append(cmdargs, command)
+
+	var cmdEnv []string
+	cmdEnv = os.Environ()
+	cmdEnv = append(cmdEnv, env...)
+
+	// Setup the reader that will read the output from the command.
+	pr, pw, err := os.Pipe()
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize pipe for output: %s", err)
+	}
+
+	// Setup the command
+	cmd := exec.Command(cmdargs[0], cmdargs[1:]...)
+	cmd.Stderr = pw
+	cmd.Stdout = pw
+	cmd.Env = cmdEnv
+
+	output := NewBuffer(DefaultInitialBufSize, p.MaxBufSize)
+
+	var readErr error
+	read := make(chan bool, 1)
+	go func() {
```

Comment:
Can we move this goroutine to a function?

Reply:
Sure, I cleaned up the goroutine logic.